### PR TITLE
[custom channels]: generate unique script keys for HTLCs

### PR DESCRIPTION
**Pull request by guggero**
Fixes the LiT integration tests that were broken after merging https://github.com/lightninglabs/taproot-assets/pull/1181.
But that basically showed us that colliding script keys for MPP shard HTLCs are actually a problem.

This PR tweaks each script key with the HTLC index, making them unique again, even if they have the same payment hash and timeout (which results in the same BTC level taproot output key).

To make the accompanying [`litd` integration tests](https://github.com/lightninglabs/lightning-terminal/pull/908) pass, we also fix the following two bugs:
 - Calculate the sweep transaction change output index correctly and dynamically to avoid invalid exclusion proofs when there are multiple HTLCs being swept at the same time.
 - Fix an issue with `UpsertScriptKey` that wouldn't allow the tweak of a script key to be updated when it is learned after the key was already added to the DB.

### CODE REVIEW

File: tapchannel/commitment.go [hunk 1]
```diff
@@ -730,6 +730,13 @@ func CreateAllocations(chanState lnwallet.AuxChanState, ourBalance,
 				"allocation, HTLC is dust")
 		}
 
+		// To ensure uniqueness of the script key across HTLCs with the
+		// same payment hash and timeout (which would be equal
+		// otherwise), we tweak the asset level internal key of the
+		// script key with the HTLC index. We'll ONLY use this for the
+		// asset level, NOT for the BTC level.
+		tweakedTree := TweakHtlcTree(htlcTree, htlc.HtlcIndex)
```
Comment (Roasbeef) [inline tapchannel/commitment.go:738]: The other relevant area here is in the aux sweeper, when we go to derive the control block and witness scripts.

File: tapchannel/aux_leaf_signer_test.go [hunk 2]
```diff
@@ -394,3 +396,61 @@ func TestTweakHtlcTree(t *testing.T) {
...
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			tweak := AddTweakWithIndex(tc.tweak, tc.index)
+			resultBytes := tc.result.Bytes()
+			resultBigInt := new(big.Int).SetBytes(resultBytes[:])
+			tweakBigInt := new(big.Int).SetBytes(tweak)
+			// 36893488147419103230
```
Comment (Roasbeef) [inline tapchannel/aux_leaf_signer_test.go:448]: Stray comment?

File: tapchannel/aux_sweeper.go [hunk 3]
```diff
@@ -314,7 +314,7 @@ func (a *AuxSweeper) signSweepVpackets(vPackets []*tappsbt.VPacket,
 		// signature.
 		signingKey, leafToSign := applySignDescToVIn(
 			signDesc, vIn, &a.cfg.ChainParams,
-			tapscriptDesc.scriptTree.TapTweak(),
+			tapscriptDesc.scriptTree.TapTweak(), 0, false,
```
Comment (gijswijs) [inline tapchannel/aux_sweeper.go:317]: IIUC the `index` value here has no meaning since `tweakWithIndex` is set to `false`.
Wouldn't the Functional Options Pattern be more elegant for `applySignDescToVIn`? You would get something like...
```go
signingKey, leafToSign := applySignDescToVIn(
	signDesc, vIn, s.cfg.ChainParams, tapscriptRoot,
	WithIndexTweak(baseJob.HTLC.HtlcIndex)
)

```
when you *do* want to tweak, and...
```go
signingKey, leafToSign := applySignDescToVIn(
	signDesc, vIn, &a.cfg.ChainParams,
	tapscriptDesc.scriptTree.TapTweak()
)
```
when you don't.

File: tapchannel/aux_leaf_signer.go [hunk 4]
```diff
@@ -735,3 +753,102 @@ func (v *schnorrSigValidator) validateSchnorrSig(virtualTx *wire.MsgTx,
 
 	return nil
 }
+
+// HtlcIndexAsScriptKeyTweak converts the given HTLC index into a modulo N
+// scalar that can be used to tweak the internal key of the HTLC script key on
+// the asset level. The value of 1 is always added to the index to make sure
+// this value is always non-zero.
+func HtlcIndexAsScriptKeyTweak(index input.HtlcIndex) *secp256k1.ModNScalar {
```
Comment (gijswijs) [inline tapchannel/aux_leaf_signer.go:761]: All four functions below are exported, but aren't used outside of the `tapchannel` package in this PR. Do we expect the need of those functions outside of this package?

**Review by gijswijs — COMMENTED** [review]
Very nice. I did a thorough review and left a bunch of comments throughout, but no blockers AFAICT.

File: tapchannel/commitment.go (hunk 1 above)
Comment (guggero) [reply tapchannel/commitment.go:764]: Yes. I've now based this PR on your sweeper PR and added that code.

File: tapchannel/aux_leaf_signer_test.go [hunk 5]
```diff
@@ -163,3 +169,288 @@ func makeCommitSig(t *testing.T, numAssetIDs, numHTLCs int) *lnwire.CommitSig {
...
+			tweakedTree := TweakHtlcTree(tc.tree, tc.index)
+			require.Equal(
+				t, tc.result.InternalKey.SerializeCompressed(),
+				tweakedTree.InternalKey.SerializeCompressed(),
+			)
+			require.Equal(
+				t, tc.result.TaprootKey.SerializeCompressed(),
+				tweakedTree.TaprootKey.SerializeCompressed(),
+			)
+			require.Equal(t, tc.result, tweakedTree)
```
[the remark this answers is not in the document]
Comment (guggero) [reply tapchannel/aux_leaf_signer_test.go:395]: There is some overlap in the two checks. But comparing the serialized keys just made debugging way easier. And the `require.Equal()` check then still make sure the other fields (tapscript tree and root) are equal.

File: tapchannel/aux_sweeper.go (hunk 3 above)
Comment (guggero) [reply tapchannel/aux_sweeper.go]: Removed this whole code as it turned out to be wrong anyway.

File: tapchannel/aux_leaf_signer.go (hunk 4 above)
Comment (guggero) [reply tapchannel/aux_leaf_signer.go:777]: You're correct. But such stateless and potentially general purpose functions IMO don't hurt to be exported.

**Comment by guggero** [comment]
I rebased this on top of https://github.com/lightninglabs/taproot-assets/pull/1154 to verify everything works.
Only the last two commits are for this PR!

@Roasbeef I have run a full itest suite on top of this PR to validate it works.
Since we're tweaking the HTLC revocation key (which is the internal key of the script key), we currently don't need to tweak anything at the **signing** level, since revocation sweeps aren't currently implemented. Left a big TODO in there so we don't forget when we add that.
So I guess this approach is even more minimal as initially thought (at least for now).

File: tapchannel/aux_leaf_signer_test.go [hunk 6]
```diff
@@ -163,3 +169,288 @@ func makeCommitSig(t *testing.T, numAssetIDs, numHTLCs int) *lnwire.CommitSig {
...
+			require.Equal(t, tc.result, tweak)
+		})
+	}
+}
+
+// TestTweakPubKeyWithIndex tests the TweakPubKeyWithIndex function.
+func TestTweakPubKeyWithIndex(t *testing.T) {
+	randNum := test.RandInt[uint32]()
+
+	makePubKey := func(tweak uint64) *btcec.PublicKey {
```
[the remark this answers is not in the document]
Comment (gijswijs) [reply tapchannel/aux_leaf_signer_test.go]: The below test case for `TestTweakPubKeyWithIndex` would still break the test because of the overflow:
```go
{
        name:   "break test",
        pubKey: startKey,
        index:  math.MaxUint64 - randNum,
        result: pubKeyFromUint64(randNum + math.MaxUint64 - randNum + 1),
},
```
It's such a weird edge case (in a unit test no less), we might as well leave it like this. I'm just proud that I thought of an overflow to break stuff. :wink:

File: tapchannel/aux_sweeper.go (hunk 3 above)
Comment (gijswijs) [reply tapchannel/aux_sweeper.go]: IIUC you created the func `TweakedRevocationKeyRing` which tweaks the `RevocationKey` and then use that keyring in the local and remote sweeps, both for timeout and success. Since we are *in* the aux_sweeper anyway, we know for a fact we need to tweak it with the `index` so no `tweakWithIndex` flag is needed.

Have I got that right?

**Comment by gijswijs** [comment]
We are nearly there. Just two minor nits and a comment to check whether I understand the code.

**Comment by jharveyb** [comment]
> I rebased this on top of #1154 to verify everything works. Only the last two commits are for this PR!
> 

Should the base branch for PR be changed then, or does this supercede #1154 directly?

File: tapchannel/aux_sweeper.go (hunk 3 above)
Comment (guggero) [reply tapchannel/aux_sweeper.go]: Yes, we need to tweak it for every sweep of an HTLC output. Since we currently only implement timeout and success sweeps, the tweak will only affect the internal key of the control block (because we're executing a script path in both cases). Once we implement revocation, then we'll also need to tweak the signing key, as then we'll use the key path.
Not sure if that's what you meant here?

File: tapchannel/aux_leaf_signer_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (guggero) [reply tapchannel/aux_leaf_signer_test.go]: Yes, good catch! Added a simplified version of your test case in, alongside with a comment why we're currently okay with this.

File: tapchannel/aux_sweeper.go (hunk 3 above)
Comment (gijswijs) [reply tapchannel/aux_sweeper.go]: Thanks for the clarification. Yes, that's what I meant.

**Review by gijswijs — APPROVED** [review]
LGTM! :tada:

**Comment by guggero** [comment]
Thanks for the review!

Unfortunately while testing this in an itest, I ran into a bit of a road block.
Turns out the tweaking works well. BUT, the thing we tweak the keys with (`HtlcIndex`) is not stable across the two peers of the channel.
Meaning that for peer A, its `HtlcIndex` of the **outgoing** HTLCs is not necessarily exactly the same as for peer B's **incoming** HTLCs. So after a certain number of HTLCs (the number seems to be around 10 in the itest, but could be pure coincidence), the ordering is out of sync and because of that the two peers have different commitment transactions, which leads to a force close.

I think we might actually need to introduce a stable sort index as a custom records to the HTLC, so both parties can use the correct value when sorting.

**Comment by Roasbeef** [comment]
Does the in-place allocation sort need to factor in the htlc index now? 

Before this change, they were identical for a given `(amt, payhash, cltv)` HTLC tuple. Now as the internal key is tweaked by the index, they'll no longer be identical. That can cause us to mix up two allocations that have the same HTLC tuple, but differ based on their index. By default `sort.Sort` isn't stable, so the slice ordering can be changed if all the keys match.

File: tapchannel/allocation_sort.go [hunk 7]
```diff
@@ -43,7 +43,7 @@ func (s sortableAllocationSlice) Swap(i, j int) {
 }
 
 // Less is a modified BIP69 output comparison, that sorts based on value, then
-// pkScript, then CLTV value.
+// pkScript, then CLTV value and finally HtlcIndex.
 //
 // NOTE: Part of the sort.Interface interface.
 func (s sortableAllocationSlice) Less(i, j int) bool {
```
Comment (jharveyb) [inline tapchannel/allocation_sort.go:49]: non-blocking nit: AFAICT the stdlib recommendation now is to use `slices.SortFunc` instead of `sort.Sort`.

File: tapchannel/aux_leaf_signer.go [hunk 8]
```diff
@@ -751,3 +753,102 @@ func (v *schnorrSigValidator) validateSchnorrSig(virtualTx *wire.MsgTx,
...
+	tweak := HtlcIndexAsScriptKeyTweak(index)
+	secp256k1.ScalarBaseMultNonConst(tweak, &tweakTimesG)
+
+	// And finally we add the result to the key to get the tweaked key.
+	secp256k1.AddNonConst(&pubKeyJacobian, &tweakTimesG, &tweakedKey)
+
+	// Convert the tweaked key back to an affine point and create a new
+	// taproot key from it.
+	tweakedKey.ToAffine()
+	return btcec.NewPublicKey(&tweakedKey.X, &tweakedKey.Y)
```
Comment (jharveyb) [inline tapchannel/aux_leaf_signer.go:809]: From the godoc of `NewPublicKey`; does this output key need to be checked with `IsOnCurve`?

My guess is no since we know the input was on the curve.

File: tapchannel/aux_leaf_signer_test.go [hunk 9]
```diff
@@ -163,3 +179,287 @@ func makeCommitSig(t *testing.T, numAssetIDs, numHTLCs int) *lnwire.CommitSig {
...
+
+	testCases := []struct {
+		name   string
+		tree   input.ScriptTree
+		index  uint64
+		result input.ScriptTree
+	}{
+		{
+			name: "index 0",
+			tree: input.ScriptTree{
```
Comment (jharveyb) [inline tapchannel/aux_leaf_signer_test.go:327]: super-nit: `tree` is already the same here and handled as read-only, so could be deduplicated.

**Review by jharveyb — APPROVED** [review]
LGTM! Massive lift 🚀 

I think the commit message for the first commit is now wrong btw? Wrt. mentioning that the tweaking for revocation is not implemented.

File: tapchannel/aux_leaf_signer.go [hunk 10]
```diff
@@ -751,3 +753,102 @@ func (v *schnorrSigValidator) validateSchnorrSig(virtualTx *wire.MsgTx,
...
+	index++
+
+	// If we wrapped around from math.MaxUint64 to 0, we need to make sure
+	// the tweak is 1 to not cause a multiplication by zero. This should
+	// never happen, as it would mean we have more than math.MaxUint64
+	// updates in a channel, which exceeds the protocol's maximum.
+	if index == 0 {
+		return new(secp256k1.ModNScalar).SetInt(1)
+	}
+
```
Comment (ffranr) [inline tapchannel/aux_leaf_signer.go:773]: To avoid overflowing `index` we could do something like this instead:

```
	if index == math.MaxUint64 {
		return new(secp256k1.ModNScalar).SetInt(1)
	}

        index++

        ...
```

Comment (ffranr) [reply tapchannel/aux_leaf_signer.go:773]: Also, on the function name, perhaps `ScriptKeyTweakFromHtlcIndex`.

File: tapchannel/aux_leaf_signer.go [hunk 11]
```diff
@@ -751,3 +753,102 @@ func (v *schnorrSigValidator) validateSchnorrSig(virtualTx *wire.MsgTx,
...
+func TweakPubKeyWithIndex(pubKey *btcec.PublicKey,
+	index input.HtlcIndex) *btcec.PublicKey {
+
+	// We need to operate on Jacobian points, which is just a different
+	// representation of the public key that allows us to do scalar
+	// multiplication.
+	var (
+		pubKeyJacobian, tweakTimesG, tweakedKey btcec.JacobianPoint
+	)
+	pubKey.AsJacobian(&pubKeyJacobian)
```
Comment (ffranr) [inline tapchannel/aux_leaf_signer.go:796]: nit: ensure that `pubKey` is not `nil`.

File: tapchannel/aux_leaf_signer.go [hunk 12]
```diff
@@ -751,3 +753,102 @@ func (v *schnorrSigValidator) validateSchnorrSig(virtualTx *wire.MsgTx,
...
+}
+
+// TweakPubKeyWithIndex tweaks the given internal public key with the given
+// HTLC index. The tweak is derived from the index in a way that never results
+// in a zero tweak. The value of 1 is always added to the index to make sure
+// this value is always non-zero. The public key is tweaked like this:
+//
+//	tweakedKey = key + (index+1) * G
+func TweakPubKeyWithIndex(pubKey *btcec.PublicKey,
+	index input.HtlcIndex) *btcec.PublicKey {
```
Comment (ffranr) [inline tapchannel/aux_leaf_signer.go:788]: I feel as if this function and the one above should be generalised into "tweak public key by integer" function(s). Which can live in btcec eventually. None of this logic is unique to taproot assets.

File: tapchannel/allocation_sort.go (hunk 7 above)
Comment (guggero) [reply tapchannel/allocation_sort.go:49]: Makes sense. Also found a way to simplify the code quite a bit.

File: tapchannel/aux_leaf_signer.go (hunk 8 above)
Comment (guggero) [reply tapchannel/aux_leaf_signer.go:809]: Yeah, we assume the input key is on the curve.

**Comment by guggero** [comment]
> I think the commit message for the first commit is now wrong btw? Wrt. mentioning that the tweaking for revocation is not implemented.

No, it's correct. Sweeping an HTLC of a revoked commitment (old state) isn't implemented yet. In that case we would need to tweak the signing key at sign time. In any non-revocation (e.g. normal, latest-state force close), we only need to tweak the internal key when we create the output and then again when we create the control block. But no _signing_ private key needs to be tweaked as of now.

**Review by gijswijs — APPROVED** [review]
No further remarks, only questions to see if I understand it. Also the last two commits fix problems that are unrelated to HTLC script key tweak, am I right?
The change output index calculation and the upsert of tweaks are unrelated to the initial problem AFAICT.

tl;dr: LGTM! :tada:

File: tapchannel/commitment.go [hunk 13]
```diff
@@ -766,13 +780,13 @@ func CreateAllocations(chanState lnwallet.AuxChanState, ourBalance,
 			NonAssetLeaves: sibling,
 			ScriptKey: asset.ScriptKey{
 				PubKey: asset.NewScriptKey(
-					htlcTree.TaprootKey,
+					tweakedTree.TaprootKey,
```
[the remark this answers is not in the document]
Comment (guggero) [reply tapchannel/commitment.go:783]: Yes, correct. We need to make sure the sort order is stable _before_ applying any tweaks (TAP leaf or HtlcIndex). Added a comment to make this clear.

File: tapchannel/commitment.go [hunk 14]
```diff
@@ -1289,12 +1317,12 @@ func createSecondLevelHtlcAllocations(chanType channeldb.ChannelType,
 		),
 		InternalKey:    htlcTree.InternalKey,
 		NonAssetLeaves: sibling,
-		ScriptKey:      asset.NewScriptKey(htlcTree.TaprootKey),
+		ScriptKey:      asset.NewScriptKey(tweakedTree.TaprootKey),
 		SortTaprootKeyBytes: schnorr.SerializePubKey(
```
[the remark this answers is not in the document]
Comment (guggero) [reply tapchannel/commitment.go:1321]: Added a comment here too.

**Comment by guggero** [comment]
> Also the last two commits fix problems that are unrelated to HTLC script key tweak, am I right?

Correct. But those changes were required to get the integration tests passing, so they were required follow-up fixes.
Going to update the PR body with what else is fixed for future reference :+1:

File: tapchannel/allocation_sort.go [hunk 15]
```diff
@@ -14,51 +15,21 @@ import (
...
-	pkScriptCmp := bytes.Compare(
-		allocI.SortTaprootKeyBytes, allocJ.SortTaprootKeyBytes,
-	)
-	if pkScriptCmp != 0 {
-		return pkScriptCmp < 0
-	}
-
-	return allocI.CLTV < allocJ.CLTV
+	slices.SortFunc(allocations, func(i, j *Allocation) int {
+		return cmp.Or(
```
Comment (Roasbeef) [inline tapchannel/allocation_sort.go:26]: 👍

File: tapchannel/aux_sweeper.go [hunk 16]
```diff
@@ -2256,26 +2256,20 @@ func (a *AuxSweeper) sweepContracts(inputs []input.Input,
...
 		// We only need to generate an exclusion proof for the second
 		// output in the commitment transaction.
-		//
-		// TODO(roasbeef): case of no change?
 		target.ExclusionProofs = append(
 			target.ExclusionProofs, proof.TaprootProof{
-				OutputIndex:    1,
-				InternalKey:    sweepInternalKey.PubKey,
-				TapscriptProof: tsProof,
+				OutputIndex: changeOutputIndex,
```
Comment (Roasbeef) [inline tapchannel/aux_sweeper.go:2268]: Ahh good catch!

File: tapchannel/aux_sweeper.go [hunk 17]
```diff
@@ -2256,26 +2256,20 @@ func (a *AuxSweeper) sweepContracts(inputs []input.Input,
 // sweepExclusionProofGen is a helper function that generates an exclusion
 // proof for the internal key of the change output.
 func sweepExclusionProofGen(sweepInternalKey keychain.KeyDescriptor,
-) tapsend.ExclusionProofGenerator {
+	changeOutputIndex uint32) tapsend.ExclusionProofGenerator {
 
 	return func(target *proof.BaseProofParams,
 		isAnchor tapsend.IsAnchor) error {
 
-		tsProof, err := proof.CreateTapscriptProof(nil)
```
Comment (Roasbeef) [inline tapchannel/aux_sweeper.go:2243]: Why don't we need to call this any longer? Did it just amount to: 
```go
&proof.TapscriptProof{
					Bip86: true,
				},
```
?

**Review by Roasbeef — APPROVED** [review]
LGTM 🦋

File: tapdb/sqlc/queries/assets.sql [hunk 18]
```diff
@@ -862,7 +862,13 @@ INSERT INTO script_keys (
                          WHEN script_keys.declared_known IS NULL OR script_keys.declared_known = FALSE
                          THEN COALESCE(EXCLUDED.declared_known, script_keys.declared_known)
                          ELSE script_keys.declared_known
-                       END
+                       END,
+      -- If the tweak was previously unknown, we'll update to the new value.
+      tweak = CASE
```
Comment (Roasbeef) [inline tapdb/sqlc/queries/assets.sql:867]: Non-blocking, but can extend the recently added TestScriptKeyKnownUpsert test to cover this

File: tapchannel/aux_sweeper.go (hunk 17 above)
Comment (guggero) [reply tapchannel/aux_sweeper.go:2243]: Yeah, it's equivalent. But I just liked the more explicit version more, since we explicitly _know_ it's a BIP-86 here.

File: tapdb/sqlc/queries/assets.sql (hunk 18 above)
Comment (guggero) [reply tapdb/sqlc/queries/assets.sql:867]: Yeah, makes sense. Added.
